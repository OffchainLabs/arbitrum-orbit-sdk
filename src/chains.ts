import { defineChain } from 'viem';
import {
  mainnet,
  holesky,
  arbitrum as arbitrumOne,
  arbitrumNova,
  sepolia,
  arbitrumSepolia,
} from 'viem/chains';

const nitroTestnodeL1 = defineChain({
  id: 1_337,
  network: 'Nitro Testnode L1',
  name: 'nitro-testnode-l1',
  nativeCurrency: { name: 'Ether', symbol: 'ETH', decimals: 18 },
  rpcUrls: {
    default: {
      http: ['http://127.0.0.1:8545'],
    },
    public: {
      http: ['http://127.0.0.1:8545'],
    },
  },
  testnet: true,
});

const nitroTestnodeL2 = defineChain({
  id: 412_346,
  network: 'Nitro Testnode L2',
  name: 'nitro-testnode-l2',
  nativeCurrency: { name: 'Ether', symbol: 'ETH', decimals: 18 },
  rpcUrls: {
    default: {
      http: ['http://127.0.0.1:8547'],
    },
    public: {
      http: ['http://127.0.0.1:8547'],
    },
  },
  testnet: true,
});

// the L3 deployed as part of the nitro-testnode framework uses a custom fee token
const nitroTestnodeL3 = defineChain({
  id: 333333,
  network: 'Nitro Testnode L3',
  name: 'nitro-testnode-l3',
  nativeCurrency: { name: 'AppTestToken', symbol: 'APP', decimals: 18 },
  rpcUrls: {
    default: {
      http: ['http://127.0.0.1:3347'],
    },
    public: {
      http: ['http://127.0.0.1:3347'],
    },
  },
  testnet: true,
});

export const chains = [
  // mainnet
  mainnet,
  arbitrumOne,
  arbitrumNova,
  // testnet
  sepolia,
  holesky,
  arbitrumSepolia,
  // local nitro-testnode
  nitroTestnodeL1,
  nitroTestnodeL2,
  nitroTestnodeL3,
] as const;

export {
  // mainnet
  mainnet,
  arbitrumOne,
  arbitrumNova,
  // testnet
  sepolia,
  holesky,
  arbitrumSepolia,
  // local nitro-testnode
  nitroTestnodeL1,
  nitroTestnodeL2,
  nitroTestnodeL3,
};
