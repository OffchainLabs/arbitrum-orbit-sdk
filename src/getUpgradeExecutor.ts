import { Address, Chain, PublicClient, Transport, getAbiItem } from 'viem';
import { rollupAdminLogicABI } from './abi';
import { createRollupFetchTransactionHash } from './createRollupFetchTransactionHash';
import { isValidParentChainId } from './types/ParentChain';
import { arbOwnerPublic, upgradeExecutor } from './contracts';
import { UPGRADE_EXECUTOR_ROLE_ADMIN } from './upgradeExecutorEncodeFunctionData';

const AdminChangedAbi = getAbiItem({ abi: rollupAdminLogicABI, name: 'AdminChanged' });

export type GetUpgradeExecutorParams = {
  /** Address of the rollup we're getting logs from */
  rollup: Address;
};
/**
 * Address of the current upgrade executor
 */
export type GetUpgradeExecutorReturnType = Address | undefined;

/**
 * Return upgrade executor address for a parent or child chain
 *
 * Docs: https://docs.arbitrum.io/launch-orbit-chain/concepts/chain-ownership
 *
 * @param {PublicClient} publicClient - The chain Viem Public Client
 * @param {GetUpgradeExecutorParams} GetUpgradeExecutorParams {@link GetUpgradeExecutorParams}
 *
 * @returns Promise<{@link GetUpgradeExecutorReturnType}>
 *
 * @example
 * const upgradeExecutor = getUpgradeExecutor(client, {
 *   rollup: '0xc47dacfbaa80bd9d8112f4e8069482c2a3221336'
 * });
 *
 */
export async function getUpgradeExecutor<TChain extends Chain | undefined>(
  publicClient: PublicClient<Transport, TChain>,
  params?: GetUpgradeExecutorParams,
): Promise<GetUpgradeExecutorReturnType> {
  const isParentChain = isValidParentChainId(publicClient.chain?.id);
  if (isParentChain && !params?.rollup) {
    throw new Error('[getUpgradeExecutor] requires a rollup address');
  }

  // Parent chain, get the newOwner args from the last event
  if (isParentChain && params) {
    let blockNumber: bigint | 'earliest';

    try {
      const createRollupTransactionHash = await createRollupFetchTransactionHash({
        rollup: params.rollup,
        publicClient,
      });
      const receipt = await publicClient.waitForTransactionReceipt({
        hash: createRollupTransactionHash,
      });
      blockNumber = receipt.blockNumber;
    } catch (e) {
      console.warn(`[getUpgradeExecutor] ${(e as any).message}`);
      blockNumber = 'earliest';
    }

    const events = await publicClient.getLogs({
      address: params.rollup,
      /**
       * The event comes from:
       * - event AdminChanged(address previousAdmin, address newAdmin)
       * - ERC1967Upgrade
       * - DoubleLogicUUPSUpgradeable
       * - RollupAdminLogic
       *
       * see https://github.com/OffchainLabs/nitro-contracts/blob/90037b996509312ef1addb3f9352457b8a99d6a6/src/rollup/RollupAdminLogic.sol#L182
       */
      events: [AdminChangedAbi],
      fromBlock: blockNumber,
      toBlock: 'latest',
    });

    return events[events.length - 1].args.newAdmin;
  }

  // Child chain, check for all chainOwners
  const chainOwners = await publicClient.readContract({
    abi: arbOwnerPublic.abi,
    functionName: 'getAllChainOwners',
    address: arbOwnerPublic.address,
  });

  const results = await Promise.allSettled(
    chainOwners.map((chainOwner) =>
      publicClient.readContract({
        address: chainOwner,
        abi: upgradeExecutor.abi,
        functionName: 'hasRole',
        args: [UPGRADE_EXECUTOR_ROLE_ADMIN, chainOwner],
      }),
    ),
  );
  const upgradeExecutorIndex = results.findIndex(
    (p) => p.status === 'fulfilled' && p.value === true,
  );
  return chainOwners[upgradeExecutorIndex];
}
