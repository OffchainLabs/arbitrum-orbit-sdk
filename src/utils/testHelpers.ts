import { Address, PrivateKeyAccount, PublicClient } from 'viem';
import {
  CreateRollupPrepareConfigResult,
  createRollupPrepareConfig,
} from '../createRollupPrepareConfig';
import { createRollupPrepareTransactionRequest } from '../createRollupPrepareTransactionRequest';
import { prepareChainConfig } from '../prepareChainConfig';
import { generateChainId } from './generateChainId';
import {
  CreateRollupTransaction,
  createRollupPrepareTransaction,
} from '../createRollupPrepareTransaction';
import {
  CreateRollupTransactionReceipt,
  createRollupPrepareTransactionReceipt,
} from '../createRollupPrepareTransactionReceipt';
import { CoreContracts } from '../types/CoreContracts';

export type CreateTestRollupParams = {
  deployer: PrivateKeyAccount;
  batchPoster: Address;
  validators: Address[];
  publicClient: PublicClient;
};

export type CreateTestRollupResult = {
  config: CreateRollupPrepareConfigResult;
  transaction: CreateRollupTransaction;
  transactionReceipt: CreateRollupTransactionReceipt;
  coreContracts: CoreContracts;
};

export async function createTestRollup({
  deployer,
  batchPoster,
  validators,
  publicClient,
}: CreateTestRollupParams): Promise<CreateTestRollupResult> {
  // generate a random chain id
  const chainId = generateChainId();

  // create the chain config
  const chainConfig = prepareChainConfig({
    chainId,
    arbitrum: { InitialChainOwner: deployer.address },
  });

  const config = createRollupPrepareConfig({
    chainId: BigInt(chainId),
    owner: deployer.address,
    chainConfig,
  });

  // prepare the transaction for deploying the core contracts
  const request = await createRollupPrepareTransactionRequest({
    params: {
      config,
      batchPoster,
      validators,
    },
    account: deployer.address,
    publicClient,
  });

  // sign and send the transaction
  const txHash = await publicClient.sendRawTransaction({
    serializedTransaction: await deployer.signTransaction(request),
  });

  // get the transaction
  const tx = createRollupPrepareTransaction(await publicClient.getTransaction({ hash: txHash }));

  // get the transaction receipt after waiting for the transaction to complete
  const txReceipt = createRollupPrepareTransactionReceipt(
    await publicClient.waitForTransactionReceipt({ hash: txHash }),
  );

  const coreContracts = txReceipt.getCoreContracts();

  return {
    config,
    transaction: tx,
    transactionReceipt: txReceipt,
    coreContracts,
  };
}
