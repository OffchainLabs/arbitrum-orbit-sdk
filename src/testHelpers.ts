import { Address, Client, PublicClient, zeroAddress } from 'viem';
import { privateKeyToAccount, PrivateKeyAccount } from 'viem/accounts';
import { config } from 'dotenv';
import { execSync } from 'node:child_process';

import { generateChainId, sanitizePrivateKey } from './utils';
import { createRollup } from './createRollup';
import { createRollupPrepareDeploymentParamsConfig } from './createRollupPrepareDeploymentParamsConfig';
import { prepareChainConfig } from './prepareChainConfig';

config();

type PrivateKeyAccountWithPrivateKey = PrivateKeyAccount & { privateKey: `0x${string}` };
// Source: https://github.com/OffchainLabs/nitro-testnode/blob/release/scripts/accounts.ts#L28
type NitroTestNodePrivateKeyAccounts = {
  // funnel
  deployer: PrivateKeyAccountWithPrivateKey;
  // sequencer (batch poster and rollup owner are the same in nitro-testnode)
  l2RollupOwner: PrivateKeyAccountWithPrivateKey;
  // l3owner
  l3RollupOwner: PrivateKeyAccountWithPrivateKey;
  // sha256(user_token_bridge_deployer)
  l3TokenBridgeDeployer: PrivateKeyAccountWithPrivateKey;
  // l2 token bridge deployer
  l2TokenBridgeDeployer: PrivateKeyAccountWithPrivateKey;
  // l3 token bridge deployer which holds custom gas token
};

export function getNitroTestnodePrivateKeyAccounts(): NitroTestNodePrivateKeyAccounts {
  if (
    typeof process.env.NITRO_TESTNODE_DEPLOYER_PRIVATE_KEY === 'undefined' ||
    typeof process.env.NITRO_TESTNODE_L2_ROLLUP_OWNER_PRIVATE_KEY === 'undefined' ||
    typeof process.env.NITRO_TESTNODE_L3_ROLLUP_OWNER_PRIVATE_KEY === 'undefined' ||
    typeof process.env.NITRO_TESTNODE_L2_TOKEN_BRIDGE_DEPLOYER_PRIVATE_KEY === 'undefined' ||
    typeof process.env.NITRO_TESTNODE_L3_TOKEN_BRIDGE_DEPLOYER_PRIVATE_KEY === 'undefined'
  ) {
    throw Error(
      `required env variables: NITRO_TESTNODE_DEPLOYER_PRIVATE_KEY, NITRO_TESTNODE_L2_ROLLUP_OWNER_PRIVATE_KEY,
      NITRO_TESTNODE_L2_TOKEN_BRIDGE_DEPLOYER_PRIVATE_KEY, NITRO_TESTNODE_L3_ROLLUP_OWNER_PRIVATE_KEY,
      NITRO_TESTNODE_L3_TOKEN_BRIDGE_DEPLOYER_PRIVATE_KEY`,
    );
  }

  const deployerPrivateKey = sanitizePrivateKey(process.env.NITRO_TESTNODE_DEPLOYER_PRIVATE_KEY);
  const l2RollupOwnerPrivateKey = sanitizePrivateKey(
    process.env.NITRO_TESTNODE_L2_ROLLUP_OWNER_PRIVATE_KEY,
  );
  const l2TokenBridgeDeployerPrivateKey = sanitizePrivateKey(
    process.env.NITRO_TESTNODE_L2_TOKEN_BRIDGE_DEPLOYER_PRIVATE_KEY,
  );
  const l3RollupOwnerPrivateKey = sanitizePrivateKey(
    process.env.NITRO_TESTNODE_L3_ROLLUP_OWNER_PRIVATE_KEY,
  );
  const l3TokenBridgeDeployerPrivateKey = sanitizePrivateKey(
    process.env.NITRO_TESTNODE_L3_TOKEN_BRIDGE_DEPLOYER_PRIVATE_KEY,
  );

  return {
    deployer: { ...privateKeyToAccount(deployerPrivateKey), privateKey: deployerPrivateKey },
    l2RollupOwner: {
      ...privateKeyToAccount(l2RollupOwnerPrivateKey),
      privateKey: l2RollupOwnerPrivateKey,
    },
    l2TokenBridgeDeployer: {
      ...privateKeyToAccount(l2TokenBridgeDeployerPrivateKey),
      privateKey: l2TokenBridgeDeployerPrivateKey,
    },
    l3RollupOwner: {
      ...privateKeyToAccount(l3RollupOwnerPrivateKey),
      privateKey: l3RollupOwnerPrivateKey,
    },
    l3TokenBridgeDeployer: {
      ...privateKeyToAccount(l3TokenBridgeDeployerPrivateKey),
      privateKey: l3TokenBridgeDeployerPrivateKey,
    },
  };
}

type TestnodeInformation = {
  bridge: Address;
  rollup: Address;
  sequencerInbox: Address;
  l3SequencerInbox: Address;
  l3Bridge: Address;
  batchPoster: Address;
  l3BatchPoster: Address;
  l3UpgradeExecutor: Address;
  l3Rollup: `0x${string}`;
  l3NativeToken: `0x${string}`;
};

export function getInformationFromTestnode(): TestnodeInformation {
  const containers = [
    'nitro_sequencer_1',
    'nitro-sequencer-1',
    'nitro-testnode-sequencer-1',
    'nitro-testnode_sequencer_1',
  ];

  for (const container of containers) {
    try {
      const deploymentJson = JSON.parse(
        execSync('docker exec ' + container + ' cat /config/deployment.json').toString(),
      );

      const l3DeploymentJson = JSON.parse(
        execSync('docker exec ' + container + ' cat /config/l3deployment.json').toString(),
      );

      const sequencerConfig = JSON.parse(
        execSync('docker exec ' + container + ' cat /config/sequencer_config.json').toString(),
      );

      const l3SequencerConfig = JSON.parse(
        execSync('docker exec ' + container + ' cat /config/l3node_config.json').toString(),
      );

      return {
        bridge: deploymentJson['bridge'],
        rollup: deploymentJson['rollup'],
        sequencerInbox: deploymentJson['sequencer-inbox'],
        batchPoster: sequencerConfig.node['batch-poster']['parent-chain-wallet'].account,
        l3Bridge: l3DeploymentJson['bridge'],
        l3Rollup: l3DeploymentJson['rollup'],
        l3SequencerInbox: l3DeploymentJson['sequencer-inbox'],
        l3NativeToken: l3DeploymentJson['native-token'],
        l3BatchPoster: l3SequencerConfig.node['batch-poster']['parent-chain-wallet'].account,
        l3UpgradeExecutor: l3DeploymentJson['upgrade-executor'],
      };
    } catch {
      // empty on purpose
    }
  }

  throw new Error('nitro-testnode sequencer not found');
}

export async function createRollupHelper({
  deployer,
  batchPoster,
  validators,
  nativeToken = zeroAddress,
  client,
}: {
  deployer: PrivateKeyAccountWithPrivateKey;
  batchPoster: Address;
  validators: [Address];
  nativeToken: Address;
  client: PublicClient;
}) {
  const chainId = generateChainId();

  const createRollupConfig = createRollupPrepareDeploymentParamsConfig(client, {
    chainId: BigInt(chainId),
    owner: deployer.address,
    chainConfig: prepareChainConfig({
      chainId,
      arbitrum: {
        InitialChainOwner: deployer.address,
        DataAvailabilityCommittee: true,
      },
    }),
  });

  const createRollupInformation = await createRollup({
    params: {
      config: createRollupConfig,
      batchPoster,
      validators,
      nativeToken,
    },
    account: deployer,
    parentChainPublicClient: client,
  });

  // create test rollup with ETH as gas token
  return {
    createRollupConfig,
    createRollupInformation,
  };
}
