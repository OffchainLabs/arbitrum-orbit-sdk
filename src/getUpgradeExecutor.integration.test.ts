import { describe, it, expect } from 'vitest';
import { createPublicClient, http } from 'viem';

import { nitroTestnodeL2, nitroTestnodeL3 } from './chains';
import { getInformationFromTestnode } from './testHelpers';
import { getUpgradeExecutor } from './getUpgradeExecutor';

const { l3UpgradeExecutor, l3Rollup } = getInformationFromTestnode();

// Tests can be enabled once we run one node per integration test
describe('successfully get upgrade executor', () => {
  it('from parent chain', async () => {
    const parentChainClient = createPublicClient({
      chain: nitroTestnodeL2,
      transport: http(),
    });

    const upgradeExecutor = await getUpgradeExecutor(parentChainClient, {
      rollup: l3Rollup,
    });
    expect(upgradeExecutor?.toLowerCase()).toEqual(l3UpgradeExecutor);
  });

  it('from child chain', async () => {
    const childChainClient = createPublicClient({
      chain: nitroTestnodeL3,
      transport: http(),
    });
    const upgradeExecutor = await getUpgradeExecutor(childChainClient);
    expect(upgradeExecutor).toMatchSnapshot();
  });
});
