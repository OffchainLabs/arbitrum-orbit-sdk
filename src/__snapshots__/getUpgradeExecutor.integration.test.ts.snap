// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`successfully get upgrade executor > from child chain 1`] = `"0x24198F8A339cd3C47AEa3A764A20d2dDaB4D1b5b"`;
